### PR TITLE
Processor: `HarvestHolderRewards`

### PR DESCRIPTION
**Pull request by febo**
This PR adds the processor for the `HarvestHolderRewards` instruction.

### CODE REVIEW

File: program/src/processor/harvest_holder_rewards.rs [hunk 1]
```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
...
+    require!(
+        ctx.accounts.holder_rewards.owner == &paladin_rewards_program_client::ID,
+        ProgramError::InvalidAccountOwner,
+        "holder rewards",
+    );
+
+    let (derivation, _) = Pubkey::find_program_address(
+        &["holder".as_bytes(), ctx.accounts.vault.key.as_ref()],
+        &paladin_rewards_program_client::ID,
+    );
```
Comment (buffalojoec) [inline program/src/processor/harvest_holder_rewards.rs]: I _swore_ we added helpers for this, what happened to those? I can't find them in the Rust client anymore.

File: program/src/state/mod.rs [hunk 2]
```diff
@@ -57,3 +66,31 @@ pub fn get_stake_pda_signer_seeds<'a>(
         bump_seed,
     ]
 }
+
+/// Calculate the eligible rewards for a given token account balance.
+///
+/// This uses the same logic as the Rewards program.
+pub fn calculate_eligible_rewards(
```
Comment (buffalojoec) [inline program/src/state/mod.rs:73]: Lmk if you think we should somehow land this in the Rewards program client so we don't have to reimplement here.

File: program/src/processor/harvest_holder_rewards.rs [hunk 3]
```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
...
+    if rewards != 0 {
+        // update the last seen holder rewards
+        stake.set_last_seen_holder_rewards_per_token(
+            holder_rewards.last_accumulated_rewards_per_token,
+        );
+
+        // Rewards are stored on the `vault` token account. We need to first withdraw the excess lamports
+        // from the vault token account to the vault authority account. Then transfer the rewards amount
+        // to the destination account and send the remaining excess lamports back to the vault token
+        // account.
```
Comment (buffalojoec) [inline program/src/processor/harvest_holder_rewards.rs:195]: Wouldn't it be easier to withdraw excess lamports to the destination account, then send the remaining excess lamports back to the vault token account? That would save a jump in between.

File: program/src/processor/harvest_holder_rewards.rs [hunk 4]
```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
...
+            ctx.accounts.vault.key,
+            ctx.accounts.vault_authority.key,
+            ctx.accounts.vault_authority.key,
+            &[],
+        )?;
+
+        let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+        // Stores the current lamports of the vault authority account before the withdraw so
+        // we can calculate the amount of rewards withdrawn.
+        let current_lamports = ctx.accounts.vault_authority.lamports();
```
Comment (buffalojoec) [inline program/src/processor/harvest_holder_rewards.rs]: nit: any chance we can make this a bit more clear? It becomes hard to follow after a few steps below.

```suggestion
        let vault_authority_starting_lamports = ctx.accounts.vault_authority.lamports();
```

File: program/src/processor/harvest_holder_rewards.rs [hunk 5]
```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
...
+        // If the withdraw does not result in enough lamports to cover the rewards, only
+        // harvest the available lamports. This should never happen, but the check is a
+        // failsafe.
+        let rewards = std::cmp::min(
+            rewards,
+            ctx.accounts
+                .vault_authority
+                .lamports()
+                .saturating_sub(current_lamports),
+        );
```
Comment (buffalojoec) [inline program/src/processor/harvest_holder_rewards.rs]: What happens if this resolves to zero? In the case of no excess lamports on the vault, for example.

File: clients/rust/tests/harvest_holder_rewards.rs [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+        context.last_blockhash,
+    );
+    context.banks_client.process_transaction(tx).await.unwrap();
+
+    // Then the destination account has the rewards.
+
+    let account = get_account!(context, destination);
+    assert_eq!(account.lamports, 2_000_000_000);
+
+    // And there should be rewards left on the vault account.
```
Comment (buffalojoec) [inline clients/rust/tests/harvest_holder_rewards.rs:133]: Maybe just out of an abundance of caution, check that the vault authority account didn't hang onto any lingering lamports?

File: clients/rust/tests/harvest_holder_rewards.rs [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+    );
+
+    // And the stake account last seen holder rewards per token is update.
+
+    let account = get_account!(context, stake_manager.stake);
+    let stake_account = Stake::from_bytes(account.data.as_ref()).unwrap();
+    assert_eq!(
+        stake_account.last_seen_holder_rewards_per_token,
+        40_000_000 * 1_000_000_000
+    );
```
Comment (buffalojoec) [inline clients/rust/tests/harvest_holder_rewards.rs:407]: This gets set from the first harvest, though, right? The second one should see zero rewards eligible and terminate early?

**Review by buffalojoec — COMMENTED** [review]
Looking great so far!!

File: program/src/processor/harvest_holder_rewards.rs (hunk 1 above)
Comment (febo) [reply program/src/processor/harvest_holder_rewards.rs]: It is there, just updated to use it.

File: program/src/processor/harvest_holder_rewards.rs (hunk 3 above)
Comment (febo) [reply program/src/processor/harvest_holder_rewards.rs:195]: The destination is not a signer, so I don't think we can do that.

File: program/src/processor/harvest_holder_rewards.rs (hunk 5 above)
Comment (febo) [reply program/src/processor/harvest_holder_rewards.rs]: It will invoke the `transfer` with `0`, which essentially will do nothing. In practice, this should not happen since we checked that we have rewards to distribute. Perhaps we can add a log message?

File: program/src/state/mod.rs (hunk 2 above)
Comment (joncinque) [reply program/src/state/mod.rs:73]: +1 on this, can be done in a future PR though

File: clients/rust/tests/harvest_holder_rewards.rs (hunk 7 above)
Comment (febo) [reply clients/rust/tests/harvest_holder_rewards.rs:407]: Yes, the test is asserting that the value of `last_seen_holder_rewards_per_token` did not change from the first harvest.

File: program/src/state/mod.rs (hunk 2 above)
Comment (febo) [reply program/src/state/mod.rs:73]: Yeah, might be a good idea.

File: program/src/processor/harvest_holder_rewards.rs [hunk 8]
```diff
@@ -21,11 +36,228 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
...
+                ctx.accounts.vault_authority.key,
+                ctx.accounts.destination.key,
+                rewards,
+            ),
+            &[
+                ctx.accounts.vault_authority.clone(),
+                ctx.accounts.destination.clone(),
+            ],
+            &[&signer_seeds],
+        )?;
```
Comment (joncinque) [inline program/src/processor/harvest_holder_rewards.rs:244]: I think it's OK either way, but we could add a check that `vault_authority != destination`

File: program/src/processor/harvest_holder_rewards.rs (hunk 3 above)
Comment (joncinque) [reply program/src/processor/harvest_holder_rewards.rs:195]: Yeah, since `withdraw_excess_lamports` takes everything, we unfortunately need to do this extra step

**Review by joncinque — APPROVED** [review]
Looks great to me!

File: program/src/processor/harvest_holder_rewards.rs (hunk 8 above)
Comment (febo) [reply program/src/processor/harvest_holder_rewards.rs:244]: Done - added a test for it as well.

File: program/src/processor/harvest_holder_rewards.rs (hunk 3 above)
Comment (buffalojoec) [reply program/src/processor/harvest_holder_rewards.rs:195]: > The destination is not a signer, so I don't think we can do that.

Ohhh right I see, you can't send from it.

**Review by buffalojoec — APPROVED** [review]
